src/lem/mod.rs

```diff
@@ -92,6 +93,105 @@ pub struct Func {
 #[derive(Debug, PartialEq, Clone, Eq, Hash)]
 pub struct Var(AString);
 
+/// LEM tags
+#[derive(Copy, Debug, PartialEq, Clone, Eq, Hash)]
+pub enum Tag {
+    Expr(ExprTag),
+    Cont(ContTag),
+    Op1(Op1),
+    Op2(Op2),
+    Ctrl(CtrlTag),
+}
+
+#[derive(Copy, Debug, PartialEq, Clone, Eq, Hash)]
+pub enum CtrlTag {
+    Return,
+    MakeThunk,
+    ApplyContinuation,
+    Error,
+}
+
+impl Tag {
+    #[inline]
+    pub fn to_field<F: LurkField>(self) -> F {
+        use Tag::*;
+        match self {
+            Expr(tag) => tag.to_field(),
+            Cont(tag) => tag.to_field(),
+            Op1(tag) => tag.to_field(),
+            Op2(tag) => tag.to_field(),
+            Ctrl(tag) => tag.to_field(),
+        }
+    }
+}
+
+impl CtrlTag {
+    #[inline]
+    fn to_field<F: LurkField>(self) -> F {
+        F::from(self as u64)
+    }
+}
+
+impl std::fmt::Display for CtrlTag {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::Return => write!(f, "return#"),
+            Self::ApplyContinuation => write!(f, "apply-cont#"),
+            Self::MakeThunk => write!(f, "make-thunk#"),
+            Self::Error => write!(f, "error#"),
+        }
+    }
+}
+
+impl std::fmt::Display for Tag {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        use Tag::*;
+        match self {
+            Expr(tag) => write!(f, "expr.{}", tag),
+            Cont(tag) => write!(f, "cont.{}", tag),
+            Op1(tag) => write!(f, "op1.{}", tag),
+            Op2(tag) => write!(f, "op2.{}", tag),
+            Ctrl(tag) => write!(f, "ctrl.{}", tag),
+        }
+    }
+}
+
+/// LEM literals
+#[derive(Debug, PartialEq, Clone, Eq, Hash)]
+pub enum Lit {
+    // TODO maybe it should be a LurkField instead of u64
+    Num(u64),
```

Comment:
If the idea is that any literal `Num` can be represented, then yes… Of course there is also `lurk::num::Num`, which allows either representation so you don't pay for a field element when a `U64` will do.